File: mocks/fruit_picker_mock.py
"""
Mock Fruit Picker Service
Simulates a robot fruit picker with realistic delays and occasional failures.
"""
import random
import uuid
import time
from datetime import datetime, timedelta
from fastapi import FastAPI
import uvicorn

app = FastAPI(title="Fruit Picker Mock Service")

jobs = {}


@app.post("/api/v1/pick-fruit")
async def pick_fruit(request: dict):
    if random.random() < 0.1:
        from fastapi import HTTPException
        raise HTTPException(status_code=503, detail="Picker robot unavailable")

    job_id = f"pick_{uuid.uuid4().hex[:8]}"
    delay_seconds = random.randint(30, 60)
    completion_time = datetime.utcnow() + timedelta(seconds=delay_seconds)

    jobs[job_id] = {
        "jobId": job_id,
        "status": "IN_PROGRESS",
        "fruitType": request.get("fruitType"),
        "quantity": request.get("quantity"),
        "startedAt": datetime.utcnow().isoformat() + "Z",
        "estimatedCompletion": completion_time.isoformat() + "Z",
    }

    return {
        "jobId": job_id,
        "estimatedCompletion": completion_time.isoformat() + "Z",
    }


@app.get("/api/v1/jobs/{job_id}")
async def get_job(job_id: str):
    if job_id not in jobs:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail=f"Job {job_id} not found")

    job = jobs[job_id]

    if job["status"] == "IN_PROGRESS":
        start = datetime.fromisoformat(job["startedAt"].replace("Z", ""))
        elapsed = (datetime.utcnow() - start).total_seconds()
        if elapsed > 5:
            job["status"] = "COMPLETED"
            job["fruits"] = [
                {"type": job["fruitType"], "quantity": job["quantity"], "quality": "premium"}
            ]

    return job


@app.get("/health")
async def health():
    return {"status": "ok", "service": "fruit-picker-mock"}


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8081)

File: mocks/fruit_picker_mock.php
<?php

/**
 * Mock Fruit Picker Service
 *
 * Simulates the robot fruit picker with realistic delays and occasional failures.
 * Run with: php -S 0.0.0.0:8081 mocks/fruit_picker_mock.php
 */

$jobs = [];

$uri = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);
$method = $_SERVER['REQUEST_METHOD'];

header('Content-Type: application/json');

if ($method === 'POST' && $uri === '/api/v1/pick-fruit') {
    $body = json_decode(file_get_contents('php://input'), true);

    // Simulate occasional failures
    if (rand(1, 10) === 1) {
        http_response_code(503);
        echo json_encode(['error' => 'SERVICE_UNAVAILABLE', 'message' => 'Picker arm malfunction']);
        exit;
    }

    $jobId = 'pick_' . bin2hex(random_bytes(4));
    $completionTime = time() + rand(30, 60);

    // Store job in temp file (stateless mock workaround)
    $jobFile = sys_get_temp_dir() . '/pie_shop_job_' . $jobId . '.json';
    file_put_contents($jobFile, json_encode([
        'jobId' => $jobId,
        'status' => 'IN_PROGRESS',
        'fruitType' => $body['fruitType'] ?? 'apple',
        'quantity' => $body['quantity'] ?? 6,
        'completionTime' => $completionTime,
    ]));

    http_response_code(202);
    echo json_encode([
        'jobId' => $jobId,
        'estimatedCompletion' => date('c', $completionTime),
    ]);
    exit;
}

if ($method === 'GET' && preg_match('#^/api/v1/jobs/(.+)$#', $uri, $matches)) {
    $jobId = $matches[1];
    $jobFile = sys_get_temp_dir() . '/pie_shop_job_' . $jobId . '.json';

    if (!file_exists($jobFile)) {
        http_response_code(404);
        echo json_encode(['error' => 'JOB_NOT_FOUND']);
        exit;
    }

    $job = json_decode(file_get_contents($jobFile), true);

    if (time() >= $job['completionTime']) {
        $job['status'] = 'COMPLETED';
        $job['fruits'] = array_fill(0, $job['quantity'], ['type' => $job['fruitType'], 'quality' => 'premium']);
        file_put_contents($jobFile, json_encode($job));
    }

    echo json_encode($job);
    exit;
}

http_response_code(404);
echo json_encode(['error' => 'NOT_FOUND']);
